plugins/cpu/fu-cpu-device.c

```diff
@@ -95,12 +109,40 @@ fu_cpu_device_convert_vendor (const gchar *vendor)
 	return vendor;
 }
 
+static gboolean
+fu_cpu_is_updatable (const gchar *ucode_dir)
+{
+	g_autofree gchar *sysfs_ucode_reload = fu_common_get_path (FU_PATH_KIND_MICROCODE_RELOAD);
+
+	/* This path may not be present on older kernels and won't be supported */
+        if (g_access (sysfs_ucode_reload, R_OK|W_OK) < 0)
+                return FALSE;
+
+	if (ucode_dir != NULL) {
+		if (g_access (ucode_dir, R_OK|W_OK) < 0)
+			return FALSE;
+	} else {
+		 if (g_access ("/lib/firmware", R_OK|W_OK) < 0)
```

Comment:
> @superm1 superm1 5 days ago Collaborator
> So an important question - if you change FU_PATH_KIND_FIRMWARE_SEARCH does the /sys/devices/system/cpu/microcode/reload still work? Because then that means you don't really need write access to /lib/firmware, and this could potentially work in an atomic OS as well. You can can check if write is OK, and if not, fall back to setting that path.

The late load works for both debian and fedora systems for a custom firmware path (e.g. /var/lib/firmware). However, the early load (via initrd update) only works with dracut (via --fwdir). For debian update-initramfs, we couldn't find a way for the tool to look for the custom firmware path.

Reply:
@hughsie can you bring some Fedora atomic guys here to think about this?  Part of it turns into a tug of war on who gets to put what in the initramfs and it would be good to have their perspective.

The other thing that comes to mind is that if the initramfs is signed/measured as part of the boot fwupd might possibly be breaking that boot chain by writing into /lib/firmware.  For those reasons I think at a minimum it needs to at least be a configurable behavior to writing to `/lib/firmware` or not.

Reply:
@cgwalters can you help with the right person? the backstory here is that Intel want to regenerate the initramfs with the new CPU microcode included.

Reply:
(Tangential but important: this issue isn't specific to ostree/"Fedora Atomic" etc. - it applies to anyone who is doing "image based updates", which is a lot of them!  In ostree world we strive to avoid having things be somehow specific to ostree or Fedora)

Is the idea that CPU microcode would ship in both `microcode_ctl` *and* fwupd?  Which would win?  I didn't see this mentioned in the PR description and I skimmed the code and didn't see it there either.

Now, related to the "generic image based updates"...not all of them will even have a mechanism to regenerate the initramfs client side.  One approach is to pass multiple initramfs images (kernel will then create "last one wins" semantics), except that doesn't work for microcode because we want it updated as nearly the very first thing before any userspace code runs....

And more broadly this would be the first fwupd plugin that somehow interacts with the operating system (kernel/initramfs) right?   That's a new nontrivial "layer crossing".

I guess my vote here is that anyone who wants to ship this should *not* ship `microcode_ctl` and instead fwupd is run as part of the build system to download+inject the current-at-that-time version or so?

And honestly it just hurts my head to imagine switching to this for current Fedora CoreOS and similar systems so if we can just turn it off there until we know it at least doesn't conflict with `microcode_ctl` that'd be good.   If we wanted to enable this client side fwupd would have to learn about e.g. `rpm-ostree initramfs --enable` etc. which again is a huge layer crossing I'd like to avoid unless the advantages are compelling.

Reply:
To further illustrate the layer crossing here, ignore ostree/etc for a second and let's talk about a traditional apt/yum system.  At least in traditional yum Fedora for example, today the initramfs is regenerated in `%posttrans` as part of the `yum` process.  But nothing today would serialize `fwupd update` with `yum update` so if `fwupd update` suddenly wanted to regenerate the initramfs too, there's going to be races there.   (rpm-ostree is a daemon that only allows one transaction at once, so while you can do client side mutation we completely avoid these types of issues.  The kernel/initramfs and `/usr` filesystem tree are owned by it, period)

Reply:
If it helps, it doesn't give me the _warm-and-fuzzies_ either. I think ideally we just want an ESRT table entry and a way to flash the microcode to the actual SPI device in some way (e.g. like an actual firmware update), but I believe that's only available far into the future.

Reply:
>(Tangential but important: this issue isn't specific to ostree/"Fedora Atomic" etc. - it applies to anyone who is doing "image based updates", which is a lot of them! In ostree world we strive to avoid having things be somehow specific to ostree or Fedora)

My apologies.  Fedora Atomic is the big one that I know about, and I associate with image based updates.

>If it helps, it doesn't give me the warm-and-fuzzies either. I think ideally we just want an ESRT table entry and a way to flash the microcode to the actual SPI device in some way (e.g. like an actual firmware update), but I believe that's only available far into the future.

OEMs own the rollout of things that go to the SPI.  So you're not going to get an ESRT entry that is across the industry.
OEMs can choose to add microcode ESRT entries if they want (I know Dell does this).

Reply:
I thought that the lack of ability to make persistent microcode updates (at least for Intel) was intentional because it creates another place for someone who compromises a system to add a persistent backdoor in full control over the host CPU from early bootup.

Reply:
I understand fwupd updating initramfs could be a problem. An alternate approach could be to only do the late load (via kernel interface) and to make it persistent also do it at boot (in cpu plugin init).

Reply:
I think taking out the initramfs regeneration would be a good move; it's seeming quite controversial.

Reply:
Thinking more about it, there is a serious downside of not doing initramfs update. At boot kernel will not see any new CPU feature bits and behaviors added by the new microcode. If there are kernel features and mitigations depending on CPU feature bits they wont be enabled.

Reply:
The thing is; I don't think we're always going to be using an initramfs. If/when we migrate from using grub to using shim->kernel (either using an EFI stub or kexec) we won't have any opportunity to load the microcode before the kernel starts. I know that's probably far into the future, but it's worth thinking about as it's no secret we want to get rid of grub.

Reply:
>The thing is; I don't think we're always going to be using an initramfs. If/when we migrate from using grub to using shim->kernel (either using an EFI stub or kexec) we won't have any opportunity to load the microcode before the kernel starts. I know that's probably far into the future, but it's worth thinking about as it's no secret we want to get rid of grub.

In that scenario, there wouldn't be an initramfs appended to the back of the kernel?  I would think you have to have something if you want to access an encrypted disk with your rootfs for example..

>Thinking more about it, there is a serious downside of not doing initramfs update. At boot kernel will not see any new CPU feature bits and behaviors added by the new microcode. If there are kernel features and mitigations depending on CPU feature bits they wont be enabled.

So maybe upleveling this with a directional discussion would help figure out the right amount of effort to put here in engineering a solution.  
* Is this expecting to displace efforts from distros?  
* Is it supposed to displace efforts from OEMs?  
* Or is it supposed to just be an accelerated supplemental path to get updated microcode out?

I have real concerns with the [stability](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/24) of [accelerated ](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/42) [microcode ](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/45) [updates](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/47) that haven't been validated by other parties such as the OEM or OS vendor.

Furthermore some of those reports I linked **only** occur on [reboot](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/21).  Other reports cause [non obvious failures](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/34).

If the pipe is too easy to get this stuff rolled out, and to make it permanent (or semi-permanent in the sense that someone needs to be technical to undo it) it seems very dangerous to me given some of those reports I've linked.

Reply:
>  would think you have to have something if you want to access an encrypted disk with your rootfs for example

Yes, encrypted root gets hard, but any initramfs-type-thing would be much more minimal than the "do anything" we have now. I'd also imagine it would be static, rather than regenerated host-side.

Reply:
> > The thing is; I don't think we're always going to be using an initramfs. If/when we migrate from using grub to using shim->kernel (either using an EFI stub or kexec) we won't have any opportunity to load the microcode before the kernel starts. I know that's probably far into the future, but it's worth thinking about as it's no secret we want to get rid of grub.
> 
> In that scenario, there wouldn't be an initramfs appended to the back of the kernel? I would think you have to have something if you want to access an encrypted disk with your rootfs for example..
> 
> > Thinking more about it, there is a serious downside of not doing initramfs update. At boot kernel will not see any new CPU feature bits and behaviors added by the new microcode. If there are kernel features and mitigations depending on CPU feature bits they wont be enabled.
> 
> So maybe upleveling this with a directional discussion would help figure out the right amount of effort to put here in engineering a solution.
> 
> * Is this expecting to displace efforts from distros?
> * Is it supposed to displace efforts from OEMs?
> * Or is it supposed to just be an accelerated supplemental path to get updated microcode out?

Giving users another option to update the microcode. This is not to displace anybody, to distribute updates that haven't been validated by other parties, or to distribute updates that are not already public in the microcode git repo.

> 
> I have real concerns with the [stability](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/24) of [accelerated ](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/42) [microcode ](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/45) [updates](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/47) that haven't been validated by other parties such as the OEM or OS vendor.
> 
> Furthermore some of those reports I linked **only** occur on [reboot](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/21). Other reports cause [non obvious failures](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/issues/34).
> 
> If the pipe is too easy to get this stuff rolled out, and to make it permanent (or semi-permanent in the sense that someone needs to be technical to undo it) it seems very dangerous to me given some of those reports I've linked.

Agree. This is why the patchset didn't update the microcode in all kernels. That was on purpose to provide a backup in the case of a failure.

Reply:
> Giving users another option to update the microcode. This is not to displace anybody, ...

I think we all understood that a new option was being proposed.  The discussion is entirely about the ramifications of suddenly having two ways to do things and the interactions between them.

Reply:
We are looking into options to see if we can avoid this conflict when regenerating initramfs. Give us some time. And thanks for bringing up this point, that's a great catch.

Reply:
We are still figuring out how to take care of concurrent initramfs updates without being too intrusive, or altogether completely avoid updating initramfs. All suggestions are most welcome.

Reply:
thb, I think concentrating on ucode-to-spi is probably the route that makes long time sense -- amybe don't worry too much about the initramfs solution for now?